### PR TITLE
feat: Enable webauthn plugin for security keys

### PR DESCRIPTION
**Pull request by cpisunyer**
Following up on the earlier PR from ikimonogakari, this code will route security key requests to the webauthn plugin if the GOOGLE_AUTH_WEBAUTHN_PLUGIN env variable is set. If the variable is not set, the existing code path continues to be executed.

### CODE REVIEW

**Comment by arithmetic1728** [comment]
@ikimonogakari could you take a look as well? Thanks!

**Review by ikimonogakari — APPROVED** [review]

File: google/oauth2/challenges.py [hunk 1]
```diff
@@ -173,6 +195,55 @@ def obtain_challenge_input(self, metadata):
...
+            timeout_ms = WEBAUTHN_TIMEOUT_MS,
+            allow_credentials = allow_credentials,
+            user_verification =  'required',
+            extensions = extension
+        )
+
+        try:
+            sys.stderr.write('Please insert and touch your security key\n')
+            get_response = webauthn_handler.get(get_request)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (cpisunyer) [reply google/oauth2/challenges.py:226]: Prompt moved to `obtain_challege_input()`

File: google/oauth2/challenges.py [hunk 2]
```diff
@@ -173,6 +195,55 @@ def obtain_challenge_input(self, metadata):
...
+    def _obtain_challenge_input_webauthn(self, metadata, webauthn_handler):
+        sk = metadata["securityKey"]
+        challenges = sk["challenges"]        
+        application_id = sk["applicationId"]
+        relying_party_id = sk["relyingPartyId"]
+
+        allow_credentials = []
+        for challenge in challenges:
+            key_handle = self._urlsafe_b64recode(challenge['keyHandle'])
+            # TODO: do we need to set transports
```
[the remark this answers is not in the document]
Comment (cpisunyer) [reply google/oauth2/challenges.py]: After chat discussion, we decided to leave the value unset.

File: google/oauth2/challenges.py [hunk 3]
```diff
@@ -173,6 +192,51 @@ def obtain_challenge_input(self, metadata):
...
             return None
 
+    def _obtain_challenge_input_webauthn(self, metadata, webauthn_handler):
+        sk = metadata["securityKey"]
+        challenges = sk["challenges"]
+        application_id = sk["applicationId"]
+        relying_party_id = sk["relyingPartyId"]
+
+        allow_credentials = []
+        for challenge in challenges:
```
Comment (sai-sunder-s) [inline google/oauth2/challenges.py:202]: why are there multiple challenges for security key?

File: google/oauth2/challenges.py [hunk 4]
```diff
@@ -110,6 +117,18 @@ def is_locally_eligible(self):
 
     @_helpers.copy_docstring(ReauthChallenge)
     def obtain_challenge_input(self, metadata):
+        # Check if there is an available Webauthn Handler, if not use pyu2f
+        try:
+            factory = webauthn_handler_factory.WebauthnHandlerFactory()
+            webauthn_handler = factory.get_handler()
+            if webauthn_handler is not None:
+                sys.stderr.write("Please insert and touch your security key\n")
```
Comment (sai-sunder-s) [inline google/oauth2/challenges.py:125]: Does webauthn handle only security key and not password?

File: google/oauth2/challenges.py [hunk 5]
```diff
@@ -173,6 +192,51 @@ def obtain_challenge_input(self, metadata):
...
+        for challenge in challenges:
+            key_handle = self._urlsafe_b64recode(challenge["keyHandle"])
+            allow_credentials.append(PublicKeyCredentialDescriptor(id=key_handle))
+
+        extension = AuthenticationExtensionsClientInputs(appid=application_id)
+
+        get_request = GetRequest(
+            origin=REAUTH_ORIGIN,
+            rpid=relying_party_id,
+            challenge=self._urlsafe_b64recode(challenges[0]["challenge"]),
```
Comment (sai-sunder-s) [inline google/oauth2/challenges.py:211]: is it possible for challenges to be empty or null

File: google/oauth2/challenges.py [hunk 6]
```diff
@@ -110,6 +117,18 @@ def is_locally_eligible(self):
...
     @_helpers.copy_docstring(ReauthChallenge)
     def obtain_challenge_input(self, metadata):
+        # Check if there is an available Webauthn Handler, if not use pyu2f
+        try:
+            factory = webauthn_handler_factory.WebauthnHandlerFactory()
+            webauthn_handler = factory.get_handler()
+            if webauthn_handler is not None:
+                sys.stderr.write("Please insert and touch your security key\n")
+                return self._obtain_challenge_input_webauthn(metadata, webauthn_handler)
+        except Exception:
```
Comment (clundin25) [inline google/oauth2/challenges.py:127]: Is it possible to capture a more specific exception? What do you expect to be thrown?

File: google/oauth2/challenges.py [hunk 7]
```diff
@@ -110,6 +117,18 @@ def is_locally_eligible(self):
...
+            # Attempt pyu2f if exception in webauthn flow
+            # traceback.print_exc()
```
Comment (clundin25) [inline google/oauth2/challenges.py:129]: nit: Don't check in commented code

File: google/oauth2/challenges.py [hunk 8]
```diff
@@ -173,6 +192,51 @@ def obtain_challenge_input(self, metadata):
                 sys.stderr.write("No security key found.\n")
             return None
 
+    def _obtain_challenge_input_webauthn(self, metadata, webauthn_handler):
+        sk = metadata["securityKey"]
+        challenges = sk["challenges"]
```
Comment (clundin25) [inline google/oauth2/challenges.py:197]: Are these fields always available in the JSON?  I want to make sure it has been validated previously so we don't throw a `KeyError`

File: google/oauth2/challenges.py (hunk 3 above)
Comment (cpisunyer) [reply google/oauth2/challenges.py:202]: Some users may have more than one security key registered.

File: google/oauth2/challenges.py [hunk 9]
```diff
@@ -22,12 +22,19 @@
...
+    GetRequest,
+    PublicKeyCredentialDescriptor,
+)
 
 
 REAUTH_ORIGIN = "https://accounts.google.com"
 SAML_CHALLENGE_MESSAGE = (
     "Please run `gcloud auth login` to complete reauthentication with SAML."
 )
+WEBAUTHN_TIMEOUT_MS = 120000  # Two minute timeout
```
Comment (clundin25) [inline google/oauth2/challenges.py:37]: Why two minutes?

File: google/oauth2/challenges.py [hunk 10]
```diff
@@ -173,6 +192,51 @@ def obtain_challenge_input(self, metadata):
...
+            "clientData": get_response.response.client_data_json,
+            "authenticatorData": get_response.response.authenticator_data,
+            "signatureData": get_response.response.signature,
+            "applicationId": application_id,
+            "keyHandle": get_response.id,
+            "securityKeyReplyType": 2,
+        }
+        return {"securityKey": response}
+
+    def _urlsafe_b64recode(self, s):
```
Comment (clundin25) [inline google/oauth2/challenges.py:234]:
```suggestion
    def _unpadded_urlsafe_base64_encoding(self, s):
```

**Review by clundin25 — APPROVED** [review]

**Review by sai-sunder-s — APPROVED** [review]

**Pull request merged**